src/app/core/tasks/threadpool.nim

```diff
@@ -1,260 +1,59 @@
-import # std libs
-  atomics, json, sequtils, tables
-
 import # vendor libs
-  chronicles, chronos, json_serialization, task_runner
+  json_serialization, json, chronicles, taskpools
 
 import # status-desktop libs
-  ./common
+  ./common 
 
-export
-  chronos, common, json_serialization
+export common, json_serialization
 
 logScope:
   topics = "task-threadpool"
 
 type
   ThreadPool* = ref object
-    chanRecvFromPool: AsyncChannel[ThreadSafeString]
-    chanSendToPool: AsyncChannel[ThreadSafeString]
-    thread: Thread[PoolThreadArg]
-    size: int
-    running*: Atomic[bool]
-  PoolThreadArg = object
-    chanSendToMain: AsyncChannel[ThreadSafeString]
-    chanRecvFromMain: AsyncChannel[ThreadSafeString]
-    size: int
-  TaskThreadArg = object
-    id: int
-    chanRecvFromPool: AsyncChannel[ThreadSafeString]
-    chanSendToPool: AsyncChannel[ThreadSafeString]
-  ThreadNotification = object
-    id: int
-    notice: string
-
-# forward declarations
-proc poolThread(arg: PoolThreadArg) {.thread.}
+    pool: Taskpool
+  ThreadSafeTaskArg* = distinct cstring
 
 const MaxThreadPoolSize = 16
 
-proc init(self: ThreadPool) =
-  self.chanRecvFromPool.open()
-  self.chanSendToPool.open()
-  let arg = PoolThreadArg(
-    chanSendToMain: self.chanRecvFromPool,
-    chanRecvFromMain: self.chanSendToPool,
-    size: self.size
-  )
-  createThread(self.thread, poolThread, arg)
-  # block until we receive "ready"
-  discard $(self.chanRecvFromPool.recvSync())
+proc safe*[T: TaskArg](taskArg: T): ThreadSafeTaskArg =
+  var 
+    strArgs = taskArg.encode()
+    res = cast[cstring](allocShared(strArgs.len + 1))
```

Comment:
this is fine for the current codebase - for the future, a few things could be considered:

* it looks like `encode` is used here (which I believe uses json_serialization) while `std/json` is used to decode - these are not necessarily compatible type-for-type in their json encoding - also, where is `tptr` coming from, when decoding?
* if the encoding is changed to some binary format, `0`:s in the stream might become a problem - one could consider allocating a few extra bytes and encode the length instead of relying on null-termination just to be safe
* since the infrastructure for json is already all across the codebase, this is a good choice for now - should it ever become a performance issue, one can easily switch to a binary format here - one possibility that exists is only do `json` for types that don't support `copyMem`, something like:
```nim
when supportsCopyMem(T):
  let ptr = allocshared(sizeof(T))
  copyMem(...)
else:
  ...
```

again though, at this stage of development, json is perfectly fine.

Reply:
>  also, where is tptr coming from, when decoding?

tptr is filled by the task initiator. It's a serialized `ByteAddress` pointing to a task with this signature `Task* = proc(arg: string): void {.gcsafe, nimcall.}`